pennylane/measurements/probs.py

```diff
@@ -259,23 +259,21 @@ def _count_samples(indices, batch_size, dim):
         """Count the occurrences of sampled indices and convert them to relative
         counts in order to estimate their occurrence probability."""
         num_bins, bin_size = indices.shape[-2:]
-        if batch_size is None:
-            prob = qml.math.zeros((dim, num_bins), dtype="float64")
-            # count the basis state occurrences, and construct the probability vector for each bin
-            for b, idx in enumerate(indices):
-                basis_states, counts = qml.math.unique(idx, return_counts=True)
-                prob[basis_states, b] = counts / bin_size
+        interface = qml.math.get_deep_interface(indices)
 
-            return prob
+        def _count_samples_core(indices, dim, interface):
+            return qml.math.array(
+                [[qml.math.sum(idx == p) for p in range(dim)] for idx in indices], like=interface
+            )
```

Comment:
Could this "dense" iteration over the `dim` range become a bottleneck? :thinking: In the previous version of the code, `dim` could be very large without affecting the performance significantly, if the `indices` are sparsely populated, I think?

Reply:
Good point, let's keep the old implementation when not tracing.